src/btf.cpp

```diff
@@ -749,9 +709,119 @@ std::set<std::string> BTF::get_all_structs() const
   return struct_set;
 }
 
-int BTF::get_btf_id(const std::string &name) const
+std::set<std::string> BTF::get_all_structs() const
+{
+  std::set<std::string> structs = get_all_structs_from_btf(vmlinux_btf);
+  for (auto &mod_btf : modules_btf)
+  {
+    auto mod_structs = get_all_structs_from_btf(mod_btf.btf);
+    structs.insert(mod_structs.begin(), mod_structs.end());
+  }
+  return structs;
+}
+
+// Retrieves BTF id of the given function and the FD of the BTF object
+// containing it (FD is 0 if the function is in the vmlinux BTF)
+std::pair<int, int> BTF::get_btf_id_fd(const std::string &func) const
 {
-  return btf__find_by_name_kind(btf, name.c_str(), BTF_KIND_FUNC);
+  __s32 id = btf__find_by_name_kind(vmlinux_btf, func.c_str(), BTF_KIND_FUNC);
+  if (id >= 0)
+    return { id, 0 };
+
+  for (auto &mod_btf : modules_btf)
+  {
+    id = btf__find_by_name_kind(mod_btf.btf, func.c_str(), BTF_KIND_FUNC);
+    if (id >= 0)
+      return { id, bpf_btf_get_fd_by_id(mod_btf.id) };
```

Comment:
Isn't this leaking the FD? Who's closing it?
Overall, I feel bpftrace is long overdue for a RAII FD type.

Reply:
Yes, the FD is leaked. I wasn't sure when it is supposed to be closed - is if ok to close it right after the the BPF program is loaded?